containerd-shim/kata/create.go

```diff
@@ -0,0 +1,257 @@
+	var err error
+
+	// Checks the MUST and MUST NOT from OCI runtime specification
+	if bundlePath, err = validCreateParams(containerID, bundlePath); err != nil {
+		return "", err
+	}
+
+	ociSpec, err := oci.ParseConfigJSON(bundlePath)
+	if err != nil {
+		return "", err
+	}
+
+	containerType, err := ociSpec.ContainerType()
+	if err != nil {
+		return "", err
+	}
+
+	//In the sandbox, the containers will only
+	//use the mnt space to separate the rootfs,
+	//and to share the other namesapces with host
+	//in the sandbox, thus remove those namespaces
+	//from ocispec except networkNamespace, since
+	//it has been ignored by kata-agent in sandbox.
+
+	for _, ns := range []specs.LinuxNamespaceType{
+		specs.UserNamespace,
+		specs.UTSNamespace,
+		specs.IPCNamespace,
+		specs.PIDNamespace,
+		specs.CgroupNamespace,
+	} {
+		removeNameSpace(&ociSpec, ns)
+	}
+
+	//set the network namespace path
+	//this set will be applied to sandbox's
+	//network config and hasn't nothing to
+	//do with containers in the sandbox since
+	//networkNamesapce hasn't been ignored by
+	//kata-agent in sandbox.
+	for _, n := range ociSpec.Linux.Namespaces {
+		if n.Type != specs.NetworkNamespace {
+			continue
+		}
+
+		if n.Path == "" {
+			n.Path = netns
+		}
+	}
+
+	setFactory(ctx, runtimeConfig)
+
+	disableOutput := noNeedForOutput(detach, ociSpec.Process.Terminal)
+
+	var c vc.VCContainer
+	switch containerType {
+	case vc.PodSandbox:
+		if s.sandbox != nil {
+			return "", fmt.Errorf("cannot create another sandbox in sandbox: %s", s.sandbox.ID())
+		}
+
+		c, err = createSandbox(ctx, ociSpec, *runtimeConfig, containerID, bundlePath, disableOutput)
+		if err != nil {
+			return "", err
+		}
+		s.sandbox = c.Sandbox()
+
+	case vc.PodContainer:
+		if s.sandbox == nil {
+			return "", fmt.Errorf("BUG: Cannot start the container, since the sandbox hasn't been created")
+		}
+
+		err = createContainer(s.sandbox, ociSpec, containerID, bundlePath, disableOutput)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return containerType, nil
+}
+
+func setFactory(ctx context.Context, runtimeConfig *oci.RuntimeConfig) {
+	if runtimeConfig.FactoryConfig.Template {
+		factoryConfig := vf.Config{
+			Template: true,
+			VMConfig: vc.VMConfig{
+				HypervisorType:   runtimeConfig.HypervisorType,
+				HypervisorConfig: runtimeConfig.HypervisorConfig,
+				AgentType:        runtimeConfig.AgentType,
+				AgentConfig:      runtimeConfig.AgentConfig,
+			},
+		}
+		logrus.WithField("factory", factoryConfig).Info("load vm factory")
+		f, err := vf.NewFactory(ctx, factoryConfig, true)
+		if err != nil {
+			logrus.WithError(err).Warn("load vm factory failed, about to create new one")
+			f, err = vf.NewFactory(ctx, factoryConfig, false)
+			if err != nil {
+				logrus.WithError(err).Warn("create vm factory failed")
+			}
+		}
+		if err != nil {
+			vci.SetFactory(ctx, f)
+		}
+	}
+}
+
+var systemdKernelParam = []vc.Param{
+	{
+		Key:   "init",
+		Value: "/usr/lib/systemd/systemd",
```

Comment:
Hi @jodh-intel , since we had thought the kata shimv2 would be independent from kata runtime and placed into containerd's shimv2 directory, thus I think there is no need to split those little common codes into a a separated library from cli.